File: src/main/java/com/ceraia/arena/listeners/ArenaBlockListener.java
package com.ceraia.arena.listeners;

import com.ceraia.Double;
import com.ceraia.arena.types.Arena;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;

public class ArenaBlockListener implements Listener {

    private final Double plugin;

    public ArenaBlockListener(Double plugin) {
        this.plugin = plugin;
        Bukkit.getPluginManager().registerEvents(this, plugin);
    }

    @EventHandler
    public void onBlockPlace(BlockPlaceEvent e) {
        Arena arena = plugin.getArenaManager().getArena(e.getPlayer());
        if (arena == null) {
            return;
        }

        arena.placeBlock(e.getBlockPlaced().getLocation());
    }

    @EventHandler
    public void onBlockBreak(BlockBreakEvent e) {
        Arena arena = plugin.getArenaManager().getArena(e.getPlayer());
        if (arena == null) {
            return;
        }

        if (arena.getPlacedBlocks().contains(e.getBlock().getLocation())) {
            arena.removeBlock(e.getBlock().getLocation());
            return;
        }

        e.setCancelled(true);
    }
}

File: src/main/java/com/ceraia/modules/arenas/listeners/ArenaBlockListener.kt
package com.ceraia.modules.arenas.listeners

import com.ceraia.Ceraia
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent

class ArenaBlockListener(private val plugin: Ceraia) : Listener {
    init {
        Bukkit.getPluginManager().registerEvents(this, plugin)
    }

    @EventHandler
    fun onBlockPlace(e: BlockPlaceEvent) {
        val arena = plugin.arenaModule.arenaManager!!.getArena(e.player) ?: return

        arena.placeBlock(e.blockPlaced.location)
    }

    @EventHandler
    fun onBlockBreak(e: BlockBreakEvent) {
        val arena = plugin.arenaModule.arenaManager!!.getArena(e.player) ?: return

        if (arena.placedBlocks.contains(e.block.location)) {
            arena.removeBlock(e.block.location)
            return
        }

        e.isCancelled = true
    }
}
